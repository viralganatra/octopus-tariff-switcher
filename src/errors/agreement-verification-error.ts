import { GenericError } from './generic-error';

export class AgreementVerificationError extends GenericError {}
