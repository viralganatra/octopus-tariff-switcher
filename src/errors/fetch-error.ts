import { GenericError } from './generic-error';

export class FetchError extends GenericError {}
