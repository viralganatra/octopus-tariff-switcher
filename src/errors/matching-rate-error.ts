import { GenericError } from './generic-error';

export class MatchingRateError extends GenericError {}
