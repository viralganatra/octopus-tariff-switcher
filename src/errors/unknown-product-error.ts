import { GenericError } from './generic-error';

export class UnknownProductError extends GenericError {}
