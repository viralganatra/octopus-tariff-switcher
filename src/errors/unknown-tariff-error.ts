import { GenericError } from './generic-error';

export class UnknownTariffError extends GenericError {}
