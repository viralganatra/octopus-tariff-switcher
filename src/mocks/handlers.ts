import { http, HttpResponse, graphql } from 'msw';
import {
  accountFixture,
  productAgileFixture,
  productCosyFixture,
  productGoFixture,
  productsFixture,
  standingChargeFixture,
  telemetryFixture,
  unitRatesFixture,
} from './fixtures';

export const handlers = [
  http.get('https://api.octopus.energy/v1/products', () => {
    return HttpResponse.json(productsFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/AGILE-24-10-01/', () => {
    return HttpResponse.json(productAgileFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/COSY-22-12-08/', () => {
    return HttpResponse.json(productCosyFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/GO-VAR-22-10-14/', () => {
    return HttpResponse.json(productGoFixture);
  }),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standing-charges/',
    () => {
      return HttpResponse.json(standingChargeFixture);
    },
  ),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standard-unit-rates',
    () => {
      return HttpResponse.json(unitRatesFixture);
    },
  ),
  graphql.mutation('ObtainKrakenToken', () => {
    return HttpResponse.json({
      data: {
        obtainKrakenToken: {
          token: 'foo',
        },
      },
    });
  }),
  graphql.query('Account', () => {
    return HttpResponse.json({
      data: accountFixture,
    });
  }),
  graphql.query('smartMeterTelemetry', () => {
    return HttpResponse.json({
      data: telemetryFixture,
    });
  }),
];
