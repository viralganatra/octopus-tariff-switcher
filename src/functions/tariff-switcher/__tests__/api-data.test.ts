import { http, HttpResponse, graphql } from 'msw';
import { UnknownProductError } from '../../../errors/unknown-product-error';
import { UnknownTariffError } from '../../../errors/unknown-tariff-error';
import { accountFixture } from '../../../mocks/fixtures';
import { server } from '../../../mocks/node';
import {
  getAccountInfo,
  getOppositeTariff,
  getPotentialRatesAndStandingChargeByTariff,
  getTodaysConsumptionInHalfHourlyRates,
  getTodaysUnitRatesByTariff,
  getToken,
} from '../api-data';

describe('API Data', () => {
  beforeEach(() => {
    vi.setSystemTime(new Date(2025, 2, 3));
  });

  it('should return a token from the api or reuse it if it exists', async () => {
    const dispatchRequest = vi.fn();
    server.events.on('request:start', dispatchRequest);

    const token = await getToken();
    await getToken();
    await getToken();

    expect(token).toBe('foo');

    // @ts-ignore
    const serverRequest = dispatchRequest.mock.calls[0][0].request;

    await expect(serverRequest.json()).resolves.toMatchObject(
      expect.objectContaining({
        variables: expect.objectContaining({
          input: {
            APIKey: 'API_KEY',
          },
        }),
      }),
    );

    expect(dispatchRequest).toHaveBeenCalledExactlyOnceWith(
      expect.objectContaining({
        request: expect.objectContaining({
          url: 'https://api.octopus.energy/v1/graphql/',
        }),
      }),
    );
  });

  it('should return the opposite tariff', () => {
    expect(getOppositeTariff('Agile Octopus')).toBe('Cosy Octopus');
    expect(getOppositeTariff('Cosy Octopus')).toBe('Agile Octopus');
  });

  it('should fetch the account info when the tariff is agile', async () => {
    const dispatchRequest = vi.fn();
    server.events.on('request:start', dispatchRequest);

    const accountInfo = await getAccountInfo();

    expect(accountInfo).toEqual({
      currentTariff: 'Agile Octopus',
      regionCode: 'A',
      deviceId: '00-00-00-00-00-00-99-2F',
      currentStandingCharge: 48.7881,
    });

    // @ts-ignore
    const serverRequest = dispatchRequest.mock.calls[0][0].request;

    await expect(serverRequest.json()).resolves.toMatchObject(
      expect.objectContaining({
        variables: expect.objectContaining({
          accountNumber: 'A-123456',
        }),
      }),
    );
  });

  it('should fetch the account info when the tariff is cosy', async () => {
    const fixture = structuredClone(accountFixture);

    // @ts-ignore
    fixture.account.electricityAgreements[0].tariff = {
      tariffCode: 'E-1R-COSY-24-10-01-A',
      productCode: 'COSY-24-10-01',
      standingCharge: 100,
    };

    server.use(
      graphql.query('Account', () => {
        return HttpResponse.json({
          data: fixture,
        });
      }),
    );

    const accountInfo = await getAccountInfo();

    expect(accountInfo).toEqual({
      currentTariff: 'Cosy Octopus',
      regionCode: 'A',
      deviceId: '00-00-00-00-00-00-99-2F',
      currentStandingCharge: 100,
    });
  });

  it('should throw an error when fetching the account info if the tariff is neither cosy or agile', async () => {
    const fixture = structuredClone(accountFixture);

    // @ts-ignore
    fixture.account.electricityAgreements[0].tariff.tariffCode = 'INVALID-TARIFF';

    server.use(
      graphql.query('Account', () => {
        return HttpResponse.json({
          data: fixture,
        });
      }),
    );

    const accountInfo = () => getAccountInfo();

    await expect(accountInfo).rejects.toThrowError(UnknownTariffError);
  });

  it('should fetch todays consumption', async () => {
    const dispatchRequest = vi.fn();
    server.events.on('request:start', dispatchRequest);

    const consumption = await getTodaysConsumptionInHalfHourlyRates({
      deviceId: 'deviceId',
    });

    // @ts-ignore
    const serverRequest = dispatchRequest.mock.calls[0][0].request;

    expect(consumption).toMatchSnapshot();
    expect(serverRequest.url).toBe('https://api.octopus.energy/v1/graphql/');
    expect(serverRequest.headers.get('authorization')).toBe('foo');

    await expect(serverRequest.json()).resolves.toMatchObject(
      expect.objectContaining({
        variables: expect.objectContaining({
          deviceId: 'deviceId',
          end: '2025-03-03T23:59:59Z',
          grouping: 'HALF_HOURLY',
          start: '2025-03-03T00:30:00Z',
        }),
      }),
    );
  });

  it('should fetch todays unit rates', async () => {
    const dispatchRequest = vi.fn();
    server.events.on('request:start', dispatchRequest);

    await getTodaysUnitRatesByTariff({
      tariffCode: 'E-1R-AGILE-18-02-21-A',
      productCode: 'AGILE-18-02-21',
    });

    // @ts-ignore
    const serverRequest = dispatchRequest.mock.calls[0][0].request;

    expect(serverRequest.url).toBe(
      'https://api.octopus.energy/v1/products/E-1R-AGILE-18-02-21-A/electricity-tariffs/AGILE-18-02-21/standard-unit-rates/?period_from=2025-03-03T00:00:00Z&period_to=2025-03-03T23:59:59Z',
    );
  });

  it('should fetch the potential rates and standing charge', async () => {
    const data = await getPotentialRatesAndStandingChargeByTariff({
      tariff: 'Agile Octopus',
      regionCode: 'A',
    });

    expect(data).toMatchSnapshot();
  });

  it('should throw an error if no product is found', async () => {
    server.use(
      http.get('https://api.octopus.energy/v1/products', () => {
        return HttpResponse.json({
          count: 0,
          next: null,
          previous: null,
          results: [],
        });
      }),
    );

    const data = () =>
      getPotentialRatesAndStandingChargeByTariff({
        tariff: 'Agile Octopus',
        regionCode: 'A',
      });

    await expect(data).rejects.toThrowError(UnknownProductError);
  });
});
