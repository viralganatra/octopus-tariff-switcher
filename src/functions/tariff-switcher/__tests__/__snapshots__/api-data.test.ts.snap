// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`API Data > should accept the new agreement 1`] = `
{
  "query": "
    mutation AcceptTermsAndConditions($input: AcceptTermsAndConditionsInput!) {
      acceptTermsAndConditions(input: $input) {
        acceptedVersion
      }
    }
  ",
  "variables": {
    "input": {
      "accountNumber": "A-123456",
      "enrolmentId": "123",
      "termsVersion": {
        "versionMajor": 1,
        "versionMinor": 5,
      },
    },
  },
}
`;

exports[`API Data > should fetch consumption for a specific day 1`] = `
[
  {
    "consumptionDelta": 84,
    "readAt": "2020-03-03T00:30:00+00:00",
    "readAtMs": 1583195400000,
    "unitCostInPence": 1.5117,
  },
]
`;

exports[`API Data > should fetch the account info when the tariff is agile 1`] = `
{
  "query": "
      query Account($accountNumber: String!) {
        account(accountNumber: $accountNumber) {
          electricityAgreements(active: true) {
            validFrom
            validTo
            meterPoint {
              meters(includeInactive: false) {
                serialNumber
                smartDevices {
                  deviceId
                }
              }
              mpan
            }
            tariff {
              ... on HalfHourlyTariff {
                productCode
                tariffCode
                standingCharge
              }
            }
          }
        }
      }
    ",
  "variables": {
    "accountNumber": "A-123456",
  },
}
`;

exports[`API Data > should fetch the potential rates and standing charge for a specific day 1`] = `
{
  "potentialProductCode": "AGILE-24-10-01",
  "potentialStandingCharge": 48.7881,
  "potentialUnitRates": [
    {
      "unitCostInPence": 13.23168,
      "validFromMs": 1583272800000,
      "validToMs": 1583107200000,
      "valid_from": "2020-03-03T22:00:00Z",
      "valid_to": "2020-03-02T00:00:00Z",
    },
  ],
}
`;

exports[`API Data > should fetch the potential rates and standing charge for today 1`] = `
{
  "potentialProductCode": "AGILE-24-10-01",
  "potentialStandingCharge": 48.7881,
  "potentialUnitRates": [
    {
      "unitCostInPence": 13.23168,
      "validFromMs": 1741039200000,
      "validToMs": 1740873600000,
      "valid_from": "2025-03-03T22:00:00Z",
      "valid_to": "2025-03-02T00:00:00Z",
    },
    {
      "unitCostInPence": 26.98143,
      "validFromMs": 1741028400000,
      "validToMs": 1741039200000,
      "valid_from": "2025-03-03T19:00:00Z",
      "valid_to": "2025-03-03T22:00:00Z",
    },
    {
      "unitCostInPence": 40.47225,
      "validFromMs": 1741017600000,
      "validToMs": 1741028400000,
      "valid_from": "2025-03-03T16:00:00Z",
      "valid_to": "2025-03-03T19:00:00Z",
    },
    {
      "unitCostInPence": 13.23168,
      "validFromMs": 1741006800000,
      "validToMs": 1741017600000,
      "valid_from": "2025-03-03T13:00:00Z",
      "valid_to": "2025-03-03T16:00:00Z",
    },
    {
      "unitCostInPence": 26.98143,
      "validFromMs": 1740985200000,
      "validToMs": 1741006800000,
      "valid_from": "2025-03-03T07:00:00Z",
      "valid_to": "2025-03-03T13:00:00Z",
    },
    {
      "unitCostInPence": 13.23168,
      "validFromMs": 1740974400000,
      "validToMs": 1740985200000,
      "valid_from": "2025-03-03T04:00:00Z",
      "valid_to": "2025-03-03T07:00:00Z",
    },
    {
      "unitCostInPence": 26.98143,
      "validFromMs": 1740960000000,
      "validToMs": 1740974400000,
      "valid_from": "2025-03-03T00:00:00Z",
      "valid_to": "2025-03-03T04:00:00Z",
    },
  ],
}
`;

exports[`API Data > should fetch the terms version 1`] = `
{
  "query": "
    query TermsAndConditionsForProduct($productCode: String!) {
      termsAndConditionsForProduct(productCode: $productCode) {
        name
        version
      }
    }
  ",
  "variables": {
    "productCode": "AGILE-24-10-01",
  },
}
`;

exports[`API Data > should fetch todays consumption 1`] = `
[
  {
    "consumptionDelta": 84,
    "readAt": "2025-03-03T00:30:00+00:00",
    "readAtMs": 1740961800000,
    "unitCostInPence": 1.5117,
  },
  {
    "consumptionDelta": 71,
    "readAt": "2025-03-03T01:00:00+00:00",
    "readAtMs": 1740963600000,
    "unitCostInPence": 1.1347,
  },
  {
    "consumptionDelta": 65,
    "readAt": "2025-03-03T01:30:00+00:00",
    "readAtMs": 1740965400000,
    "unitCostInPence": 1.0033,
  },
  {
    "consumptionDelta": 65,
    "readAt": "2025-03-03T02:00:00+00:00",
    "readAtMs": 1740967200000,
    "unitCostInPence": 1.0176,
  },
  {
    "consumptionDelta": 67,
    "readAt": "2025-03-03T02:30:00+00:00",
    "readAtMs": 1740969000000,
    "unitCostInPence": 1.0341,
  },
  {
    "consumptionDelta": 64,
    "readAt": "2025-03-03T03:00:00+00:00",
    "readAtMs": 1740970800000,
    "unitCostInPence": 1.002,
  },
  {
    "consumptionDelta": 73,
    "readAt": "2025-03-03T03:30:00+00:00",
    "readAtMs": 1740972600000,
    "unitCostInPence": 1.1076,
  },
  {
    "consumptionDelta": 63,
    "readAt": "2025-03-03T04:00:00+00:00",
    "readAtMs": 1740974400000,
    "unitCostInPence": 1.153,
  },
  {
    "consumptionDelta": 74,
    "readAt": "2025-03-03T04:30:00+00:00",
    "readAtMs": 1740976200000,
    "unitCostInPence": 1.1997,
  },
  {
    "consumptionDelta": 60,
    "readAt": "2025-03-03T05:00:00+00:00",
    "readAtMs": 1740978000000,
    "unitCostInPence": 1.1699,
  },
  {
    "consumptionDelta": 77,
    "readAt": "2025-03-03T05:30:00+00:00",
    "readAtMs": 1740979800000,
    "unitCostInPence": 1.6574,
  },
  {
    "consumptionDelta": 58,
    "readAt": "2025-03-03T06:00:00+00:00",
    "readAtMs": 1740981600000,
    "unitCostInPence": 1.3118,
  },
  {
    "consumptionDelta": 86,
    "readAt": "2025-03-03T06:30:00+00:00",
    "readAtMs": 1740983400000,
    "unitCostInPence": 2.2268,
  },
  {
    "consumptionDelta": 63,
    "readAt": "2025-03-03T07:00:00+00:00",
    "readAtMs": 1740985200000,
    "unitCostInPence": 1.5975,
  },
  {
    "consumptionDelta": 145,
    "readAt": "2025-03-03T07:30:00+00:00",
    "readAtMs": 1740987000000,
    "unitCostInPence": 3.8047,
  },
  {
    "consumptionDelta": 136,
    "readAt": "2025-03-03T08:00:00+00:00",
    "readAtMs": 1740988800000,
    "unitCostInPence": 3.2744,
  },
  {
    "consumptionDelta": 74,
    "readAt": "2025-03-03T08:30:00+00:00",
    "readAtMs": 1740990600000,
    "unitCostInPence": 1.5501,
  },
  {
    "consumptionDelta": 104,
    "readAt": "2025-03-03T09:00:00+00:00",
    "readAtMs": 1740992400000,
    "unitCostInPence": 1.8651,
  },
  {
    "consumptionDelta": 81,
    "readAt": "2025-03-03T09:30:00+00:00",
    "readAtMs": 1740994200000,
    "unitCostInPence": 1.1431,
  },
  {
    "consumptionDelta": 99,
    "readAt": "2025-03-03T10:00:00+00:00",
    "readAtMs": 1740996000000,
    "unitCostInPence": 1.3472,
  },
  {
    "consumptionDelta": 37,
    "readAt": "2025-03-03T10:30:00+00:00",
    "readAtMs": 1740997800000,
    "unitCostInPence": 0.4242,
  },
]
`;

exports[`API Data > should fetch todays consumption 2`] = `
{
  "query": "
    query smartMeterTelemetry(
      $deviceId: String!,
      $start: DateTime,
      $end: DateTime,
      $grouping: TelemetryGrouping
    ) {
      smartMeterTelemetry(
        deviceId: $deviceId,
        start: $start,
        end: $end,
        grouping: $grouping
      ) {
        readAt
        consumptionDelta
        costDeltaWithTax
      }
    }
  ",
  "variables": {
    "deviceId": "deviceId",
    "end": "2025-03-03T23:59:59Z",
    "grouping": "HALF_HOURLY",
    "start": "2025-03-03T00:00:00Z",
  },
}
`;

exports[`API Data > should start the onboarding process 1`] = `
{
  "query": "
    mutation StartOnboardingProcess($input: StartSmartOnboardingProcessInput) {
      startOnboardingProcess(input: $input) {
        onboardingProcess {
          id
        }
        productEnrolment {
          id
        }
        possibleErrors {
          message
          code
        }
      }
    }
  ",
  "variables": {
    "input": {
      "accountNumber": "A-123456",
      "mpan": "mpan",
      "productCode": "COSY-24-10-01",
      "targetAgreementChangeDate": "2025-03-03",
    },
  },
}
`;
